### PR TITLE
[BUG] [Return-types] Fix param shift recipe

### PR DESCRIPTION
**Pull request by antalszava**
The following fails:
```py
qml.enable_return()
dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, 0)
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliX(1))

# This is not a correct gradient recipe, but that's not the issue
recipes = ([[1, 1, 0], [-1e5, 1, -1e5]],)
x = pnp.array(0.4, requires_grad=True)
circuit(x)
tapes, fn = qml.gradients.param_shift(circuit.tape, gradient_recipes=recipes)
print( fn(qml.execute(tapes, dev, None)))
```
The scope should be:
* Tapes with multiple returned measurements together with;
* Gradient recipes that contain unshifted terms.

**Changes**

The issue is because `r0` is not being indexed in an aux function here:
https://github.com/PennyLaneAI/pennylane/blob/21b35b6ad368729b0cb9a10071af8ca4c518538c/pennylane/gradients/parameter_shift.py#L169

### CODE REVIEW

File: tests/returntypes/test_parameter_shift_new.py [hunk 1]
```diff
@@ -57,6 +57,22 @@ class DummyOp(orig_op):
...
+    def test_qnode_custom_recipe(self):
+        """Test a custom recipe using a QNode."""
+        dev = qml.device("default.qubit", wires=2)
+
+        x = np.array(0.4, requires_grad=True)
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(x, 0)
+            qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliX(1))
+
+        # This is not a correct gradient recipe, but that's not the issue
```
Comment (dwierichs) [inline tests/returntypes/test_parameter_shift_new.py:69]:
```suggestion
        # This is not a correct gradient recipe, but this test only checks execution with an unshifted term.
```

File: tests/returntypes/test_parameter_shift_new.py [hunk 2]
```diff
@@ -57,6 +57,22 @@ class DummyOp(orig_op):
...
+        recipes = ([[-1e7, 1, 0], [1e7, 1, 1e7]],)
+        tapes, fn = qml.gradients.param_shift(tape, gradient_recipes=recipes)
```
Comment (dwierichs) [inline tests/returntypes/test_parameter_shift_new.py:71]:
```suggestion
        tapes, fn = qml.gradients.param_shift(tape, gradient_recipes=recipes)
        assert len(tapes) == 2
```

**Review by dwierichs — APPROVED** [review]
Thanks for the hotfix!
Do we want a changelog entry?

**Comment by dwierichs** [comment]
I discovered  a bug in how `r0` is determined when it is not used (-> silent bug) that surfaces in the failing test here. Will push a fix shortly.

**Comment by antalszava** [comment]
Oh, I merged before applying the suggestions :cry: 

> Do we want a changelog entry?

No need, this is a new feature being added - the bug is more of an improvement.
